### PR TITLE
Convert to TypeScript

### PR DESCRIPTION
**Pull request by reillyeon**
This change converts serial.js to TypeScript (as serial.ts), adds a build command and updates the linter configuration to support TypeScript.

### CODE REVIEW

File: serial.ts [hunk 1]
```diff
@@ -367,9 +397,9 @@ class SerialPort {
    * @return {object} The interface Object created from the WebUSB API that is
    * has the specified classNumber
    */
-  getInterfaceWithClass_(device, classNumber) {
+  private getInterfaceWithClass(device: USBDevice, classNumber: number) {
     let interfaceWithClass;
-    device.configuration.interfaces.forEach((deviceInterface) => {
+    device.configuration!.interfaces.forEach((deviceInterface) => {
```
Comment (jameshollyergoogle) [inline serial.ts]: What is the purpose of this "!"?

File: serial.ts [hunk 2]
```diff
@@ -35,29 +49,30 @@ const kAcceptableDataBits = [16, 8, 7, 6, 5];
...
 
-const kParityIndexMapping = ['none', 'odd', 'even', 'mark', 'space'];
+const kParityIndexMapping: ParityType[] =
+    ['none', 'odd', 'even', 'mark', 'space'];
 const kStopbitsIndexMapping = [1, 1.5, 2];
 
 /** a class used to control serial devices over WebUSB */
-class SerialPort {
+export class SerialPort {
+  public in: ReadableStream<Uint8Array>;
```
Comment (jameshollyergoogle) [inline serial.ts]: Any particular reason you chose the arraybuffer to be typed as Uint8Array?  Am I correct in assuming the actual serial implementation will type this as Uint8Array as well?

File: serial.ts [hunk 3]
```diff
@@ -186,79 +209,86 @@ class SerialPort {
...
-              (result) => {
-                controller.enqueue(result.data);
-                readLoop();
-              },
-              (error) => {
-                controller.error(error.toString());
-              });
-    };
-    readLoop();
+  private async readStart_(controller: ReadableStreamDefaultController) {
```
Comment (jameshollyergoogle) [inline serial.ts]: I am on board with private functions not having underscores.  However, you forgot one.

File: serial.ts (hunk 1 above)
Comment (reillyeon) [reply serial.ts]: It's an assertion that the field is non-null, which is guaranteed to be true if we've called `selectConfiguration()` however I notice that we call this method before `open()` is called. I've moved these calls into `open()` after the configuration is selected.

File: serial.ts (hunk 2 above)
Comment (reillyeon) [reply serial.ts]: This indicates that the stream is a byte stream (ArrayBuffer is much more generic). This is the type that will be used by the Serial API specification. It is also required if we want to support a BYOB (bring your own buffer) stream when those are eventually implemented in Chrome.

**Pull request merged**